harvester_e2e_tests/integrations/test_4_vm_backup_restore.py

```diff
@@ -605,9 +605,227 @@ def test_restore_replace_with_vm_shutdown_command(
+        base_vm_with_data['host_ip'], base_vm_with_data['vm_ip'] = host_ip, vm_ip
+
+        # Login to the new VM and check data is existing
+        with vm_shell_from_host(host_ip, vm_ip, base_vm_with_data['ssh_user'], pkey=pri_key) as sh:
+            endtime = datetime.now() + timedelta(seconds=wait_timeout)
+            while endtime > datetime.now():
+                out, err = sh.exec_command('cloud-init status')
+                if 'done' in out:
+                    break
+                sleep(3)
+            else:
+                raise AssertionError(
+                    f"VM {restored_vm_name} Started {wait_timeout} seconds"
+                    f", but cloud-init still in {out}"
+                )
+
+            out, err = sh.exec_command(f"cat {backup_data['path']}")
+
+        assert backup_data['content'] in out, (
+            f"cloud-init writefile failed\n"
+            f"Executed stdout: {out}\n"
+            f"Executed stderr: {err}"
+        )
+
+        # teardown: delete restored vm and volumes
+        code, data = api_client.vms.get(restored_vm_name)
+        vm_spec = api_client.vms.Spec.from_dict(data)
+        api_client.vms.delete(restored_vm_name)
+        endtime = datetime.now() + timedelta(seconds=wait_timeout)
+        while endtime > datetime.now():
+            code, data = api_client.vms.get(restored_vm_name)
+            if 404 == code:
+                break
+            sleep(3)
+        else:
+            raise AssertionError(
+                f"Failed to Delete VM({restored_vm_name}) with errors:\n"
+                f"Status({code}): {data}"
+            )
+        for vol in vm_spec.volumes:
+            vol_name = vol['volume']['persistentVolumeClaim']['claimName']
+            api_client.volumes.delete(vol_name)
+
+    @pytest.mark.dependency(depends=["TestBackupRestore::tests_backup_vm"], param=True)
+    def test_with_snapshot_restore_replace_retain_vols(
+        self, api_client, vm_shell_from_host, ssh_keypair, wait_timeout, vm_checker,
+        backup_config, base_vm_with_data
+    ):
+        unique_vm_name, backup_data = base_vm_with_data['name'], base_vm_with_data['data']
+        pub_key, pri_key = ssh_keypair
+
+        vm_snapshot_name = unique_vm_name + '-snapshot-retain'
+
+        # take vm snapshot
+        code, data = api_client.vm_snapshots.create(unique_vm_name, vm_snapshot_name)
+        assert 201 == code
+
+        deadline = datetime.now() + timedelta(seconds=wait_timeout)
+        while deadline > datetime.now():
+            code, data = api_client.vm_snapshots.get(vm_snapshot_name)
+            if data.get("status", {}).get("readyToUse"):
+                break
+            print(f"waiting for {vm_snapshot_name} to be ready")
+            sleep(3)
+        else:
+            raise AssertionError(f"timed out waiting for {vm_snapshot_name} to be ready")
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
```

Comment:
same as previous.

Reply:
Thanks for the suggestion. 

Indeed these 2 lines can be put into the previous while loop since a redundant check exists on `data.get("status", {}).get("readyToUse")`. 
Update the code accordingly.